### PR TITLE
CI Improvements - Parallelize integration tests, cache ghcup, fix tag caching

### PR DESCRIPTION
**Pull request by zlav**
# Overview

Four CI improvements across integration tests, build-all, and benchmarks workflows.

Impact
- Build times when fully cached are reliably in the single digits, as low as 5m.
- Integration tests can be more aggressively added to without adding more absolute test run time.

## 1. Parallelize integration tests

Split the single integration test job into a build job + 3 parallel test matrix jobs:
- **jvm**: Maven, Gradle, Scala, Clojure
- **compiled**: Go, Rust, Erlang, Elixir
- **scripting-and-other**: Python, Ruby, Swift, Cocoapods, Pnpm, Nuget, Carthage, LicenseScanner, Container, Reachability

Build job compiles once and populates caches. Test jobs restore caches and run their subset via `--match`. Adds a concurrency group to cancel stale in-progress runs on the same branch.

A `TEST_GROUPS` env var at the workflow level lists all match patterns. After building, a validation step runs `--dry-run` to count total tests vs matched tests and fails if any test is uncovered by a group. This prevents new tests from not being added to the matrix.

Vendor binaries and Rust release binaries are uploaded as artifacts from the build job and downloaded in test jobs, avoiding redundant `cargo build --release` (~1m) and `vendor_download.sh` (~20s) per test job.

The build job exports the release tag via job outputs so test jobs use the same tag for cache keys (avoids a race if a release happens mid-run).

**Note** This update also stops previous integration tests from running if a new commit is pushed. We have limited runner space, and I've been seeing a lot of tests in the "Waiting to run this spec" state. This should fix that.

## 2. Cache ghcup in build-all.yml

Cache `~/.ghcup/{bin,ghc,cache}` on macOS and Windows. Skips the ~500 MB GHC download on cache hit, reducing the `Setup ghc/cabal` step from ~4m to ~30s. Cache key includes both GHC and cabal versions.

## 3. Fix release tag resolution

`git describe --tags --abbrev=0` requires the tagged commit to be reachable in local history. With `fetch-depth: 2`, the tag is never reachable and always resolves to `none`. Switched to `git tag --sort=-v:refname | head -1` which lists fetched tag refs regardless of shallow clone depth.

## 4. Upgrade actions/cache to v5

Upgrade `actions/cache` from v4 to v5 across all workflows (integration tests, build-all, benchmarks). v5 uses the node22 runtime. v4 uses node20 which GitHub is deprecating (forced migration to node24 starts June 2, 2026). Self-hosted runner is on v2.332.0, above the v2.327.1 minimum.

## Testing plan

1. Pushed to branch, verified all 3 test matrix jobs pass.
2. Verified cache hits on subsequent push (exact key match, no recompilation beyond `App.Version.TH` chain).
3. Verified ghcup cache hit on macOS — `Setup ghc/cabal` dropped from ~4m to ~30s.

## Risks

- Test groups may become unbalanced as tests are added/removed. Rebalance by moving patterns between groups.
- The `--dry-run` grep pattern (`grep -c "^  "`) depends on hspec's output format. If hspec changes its dry-run output, the validation step may need updating.

## Checklist

- [x] I added tests for this PR's change (or explained in the PR description why tests don't make sense).
- [x] If this PR introduced a user-visible change, I added documentation into `docs/`.
- [x] If this PR added docs, I added links as appropriate to the user manual's ToC in `docs/README.ms` and gave consideration to how discoverable or not my documentation is.
- [x] If this change is externally visible, I updated `Changelog.md`. If this PR did not mark a release, I added my changes into an `## Unreleased` section at the top.
- [x] If I made changes to `.fossa.yml` or `fossa-deps.{json.yml}`, I updated `docs/references/files/*.schema.json` AND I have updated example files used by `fossa init` command. You may also need to update these if you have added/removed new dependency type (e.g. `pip`) or analysis target type (e.g. `poetry`).
- [x] If I made changes to a subcommand's options, I updated `docs/references/subcommands/<subcommand>.md`.

### CODE REVIEW

File: .github/workflows/build-all.yml [hunk 1]
```diff
@@ -62,14 +62,25 @@ jobs:
...
+    - uses: actions/cache@v5
+      if: ${{ !contains(matrix.os-name, 'Linux') }}
+      name: Cache ghcup
+      with:
+        path: |
+          ~/.ghcup/bin
+          ~/.ghcup/ghc/${{ matrix.ghc }}
+          ~/.ghcup/cache
+        key: ${{ matrix.os-name }}-ghcup-${{ matrix.ghc }}
+
```
Comment (zlav) [inline .github/workflows/build-all.yml:85]: Done

File: .github/workflows/integrations-test.yml [hunk 2]
```diff
@@ -98,8 +100,103 @@ jobs:
...
-        cabal test --project-file=cabal.project.ci.linux --test-show-details=direct --test-option=--times integration-tests --test-options="+RTS -N -RTS"
+        cabal update
+        MATCH_ARGS=""
+        for pattern in ${{ matrix.matches }}; do
+          MATCH_ARGS="$MATCH_ARGS --test-option=--match --test-option=$pattern"
+        done
+        cabal test --project-file=cabal.project.ci.linux \
+          --test-show-details=direct \
+          --test-option=--times \
+          $MATCH_ARGS \
```
[the remark this answers is not in the document]
Comment (zlav) [reply .github/workflows/integrations-test.yml:235]: Rejected. cabal update is required — the Hackage index isn't in the cached paths and cabal test fails
   without it.

File: .github/workflows/integrations-test.yml [hunk 3]
```diff
@@ -98,8 +96,103 @@ jobs:
...
+
+    - uses: actions/checkout@v4
+      with:
+        lfs: true
+        fetch-depth: 2
+        fetch-tags: true
+
+    - name: Get latest release tag
+      id: latest-tag
+      run: echo "tag=$(git tag --sort=-v:refname | head -1 2>/dev/null || echo none)" >> $GITHUB_OUTPUT
```
[the remark this answers is not in the document]
Comment (zlav) [reply .github/workflows/integrations-test.yml]: Edge case, but fixed

File: .github/workflows/build-all.yml [hunk 4]
```diff
@@ -54,22 +54,35 @@ jobs:
...
       with:
         lfs: true
         fetch-depth: 2
         fetch-tags: true
 
     - name: Get latest release tag
       id: latest-tag
-      run: echo "tag=$(git describe --tags --abbrev=0 2>/dev/null || echo none)" >> $GITHUB_OUTPUT
+      run: |
+        tag="$(git tag --sort=-v:refname 2>/dev/null | head -1 || true)"
```
Comment (csasarak) [inline .github/workflows/build-all.yml:66]: Why not `git describe`? It seems tailor made for this: https://git-scm.com/docs/git-describe

File: .github/workflows/build-all.yml [hunk 5]
```diff
@@ -54,22 +54,35 @@ jobs:
...
+    # Cache ghcup installations so the setup action skips downloading GHC (~500 MB).
+    - uses: actions/cache@v5
+      if: ${{ !contains(matrix.os-name, 'Linux') }}
+      name: Cache ghcup
+      with:
+        path: |
+          ${{ runner.os == 'Windows' && 'C:\ghcup\bin' || '~/.ghcup/bin' }}
+          ${{ runner.os == 'Windows' && format('C:\ghcup\ghc\{0}', matrix.ghc) || format('~/.ghcup/ghc/{0}', matrix.ghc) }}
+          ${{ runner.os == 'Windows' && 'C:\ghcup\cache' || '~/.ghcup/cache' }}
+        key: ${{ matrix.os-name }}-ghcup-${{ matrix.ghc }}-3.10.3.0
```
Comment (csasarak) [inline .github/workflows/build-all.yml:84]: What is this `3.10.3.0` at the end? The version of cabal used? I think it should be in a variable which names it.

File: .github/workflows/integrations-test.yml [hunk 6]
```diff
@@ -1,41 +1,58 @@
...
+  cancel-in-progress: true
+
+# Single source of truth for integration test groups. Used by both the
+# matrix strategy and the build job's coverage validation step.
+env:
+  TEST_MATRIX: >-
+    [
+      {"group": "jvm", "matches": "Analysis.Maven Analysis.Gradle Analysis.Scala Analysis.Clojure"},
+      {"group": "compiled", "matches": "Analysis.Go Analysis.Rust Analysis.Erlang Analysis.Elixir"},
+      {"group": "scripting-and-other", "matches": "Analysis.Python Analysis.Ruby Analysis.Swift Analysis.Cocoapods Analysis.Pnpm Analysis.Nuget Analysis.Carthage Analysis.LicenseScanner Container.Analysis Reachability.Upload"}
```
Comment (csasarak) [inline .github/workflows/integrations-test.yml:21]: Why these groups?

File: .github/workflows/integrations-test.yml [hunk 7]
```diff
@@ -1,41 +1,58 @@
...
+    ]
+
 jobs:
-  integration-test:
-    name: integration-test
+  # Build once, then run test groups in parallel.
+  # The build job populates the cabal and dist-newstyle caches.
+  # Each test job restores those caches so it doesn't need to rebuild.
```
Comment (csasarak) [inline .github/workflows/integrations-test.yml:27]: This comment just describes what the code does.

File: .github/workflows/integrations-test.yml [hunk 8]
```diff
@@ -98,8 +111,127 @@ jobs:
...
+        path: vendor-bins/
+        if-no-files-found: error
+
+    - uses: actions/upload-artifact@v5
+      with:
+        name: rust-release-bins
+        path: |
+          target/release/berkeleydb
+          target/release/millhone
+        if-no-files-found: error
```
Comment (csasarak) [inline .github/workflows/integrations-test.yml:157]: I'm not super familiar with this action, but this is how we upload artifacts to display in our release lists. I don't think we want these exposed in our list of release artifacts. Can you verify for me that they won't be? 

This is blocking for me.

**Review by csasarak — CHANGES_REQUESTED** [review]
My only blocking comment is the one asking about `upload-artifact` behavior.

**Comment by zlav** [comment]
Closing this because it was
- Sequenced with the caching PR that broke MacOS without brew
- Outdated, but can easily be revived once I add some testing for the MacOS issue
